### DISCUSSION TITLE
WithPerceptionTracking warning for nested non TCA views

### DISCUSSION THREAD

**Discussion by shannoga** [pointfreeco/swift-composable-architecture#2974] [Q&A] [answered]
Hi
We have a TCA view that is wrapped with `WithPerceptionTracking`. inside it we are looping elements from the state and creating a view for each element. This view is a simple vanilla SwiftUI view, but we do pass it a Binding from the state.
We found that we have to wrap each element View inside `WithPerceptionTracking` otherwise we got the Perception not tracking alert. 
We would love to know why isn't the Top `WithPerceptionTracking` handles the inner views tracking? 

Thanks

```swift
 public var body: some View {
        WithPerceptionTracking {
            ZStack(alignment: .center) {
                if let scaleFactor {
                    ZStack(alignment: .center) {
                        ForEach(store.surface.scaled(by: scaleFactor).elements) { element in
                            WithPerceptionTracking {
                                element.view(store: store, editingElementID: $editingElementID, scaleFactor: scaleFactor)
                                    .clipped()
                            }
                        }
                    }...
```

**Comment by mbrandonw** [comment, marked as answer]
Hi @shannoga, this is covered in the [docs](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/observationbackport#Lazy-view-closures). All _escaping_ view builder closures must have a `WithPerceptionTracking` applied. So `ZStack` doesn't need one on the inside, but `ForEach` does.

**Reply by shannoga** [reply]
Thank you for your quick reply
